rspace/src/main/scala/coop/rchain/rspace/ReplayRSpace.scala

```diff
@@ -141,26 +140,37 @@ class ReplayRSpace[F[_], C, P, E, A, R, K](store: IStore[F, C, P, A, K], branch:
               }
       } yield r
 
-//    @tailrec <- todo fix tailrec
     @SuppressWarnings(Array("org.wartremover.warts.Throw"))
     // TODO stop throwing exceptions
-    def getCommOrDataCandidates(comms: Seq[COMM]): F[Either[COMM, Seq[DataCandidate[C, R]]]] =
-      comms match {
-        case Nil =>
-          val msg = "List comms must not be empty"
-          logger.error(msg)
-          throw new IllegalArgumentException(msg)
-        case commRef :: Nil =>
-          runMatcher(commRef).map {
-            case Some(x) => Right(x)
-            case None    => Left(commRef)
-          }
-        case commRef :: rem =>
-          runMatcher(commRef).flatMap {
-            case Some(x) => Right(x).asInstanceOf[Either[COMM, Seq[DataCandidate[C, R]]]].pure[F]
-            case None    => getCommOrDataCandidates(rem)
-          }
+    def getCommOrDataCandidates(comms: Seq[COMM]): F[Either[COMM, Seq[DataCandidate[C, R]]]] = {
+      type COMMOrData  = Either[COMM, Seq[DataCandidate[C, R]]]
+      type Accumulator = (Seq[COMM], Option[COMMOrData])
+      implicit class RichAccumulator(acc: Accumulator) {
```

Comment:
Why not case class? If you mean to save allocations, I think there are no savings:
 - each tuple is an object instance afaik, exchangable for case class allocation
 - each RichAccumulator method call causes another allocation, defeating the effort to save allocations